README


This archive contains the distribution Net-Bugzilla-Kanbanize,
version 0.007:

  Bugzilla and Kanbanize sync tool

This software is Copyright (c) 2014 by Philippe M. Chiasson.

This is free software, licensed under:

  Mozilla Public License Version 2.0


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



CHANGES
==================================================
Changes from 2013-06-26 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.007 at 2014-06-26 15:24:46 +0000
------------------------------------------

  Change: a39309fc25f8e9d302e5e403d3f291a559bf3310
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-26 11:24:42 +0000

    publish to people.m.o also 

  Change: 21cd82d6c488e8d3b45735241f94bb3ed1585174
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-26 11:24:30 +0000

    docs++ 

  Change: e2d3259d083b851f0a9ae0a19bb5dae19b83b1ab
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-25 17:19:22 +0000

    add --version 

  Change: ec3c32082aa0c82bda7f2d52c684601bb6dfefe8
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-25 16:21:04 +0000

    report individual changes 

  Change: ff80462fd6d740c77c90042343ffd567501cc072
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-24 20:13:43 +0000

    report itemize sync changes if verbose 

  Change: 1eab749f616249042d5e3f0755c0ab103f0408de
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-23 12:16:20 +0000

    verbose++ 

  Change: 1c75691f4ddbdf5ee6d990e9eaa0a83aa1543a12
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-23 11:54:27 +0000

    sign git tags 

  Change: 6eb4d22834a24333fa6f1dbc4fd2da19ae3cb01c
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-23 11:39:57 +0000

    ptidy++ 

  Change: 0800aeada25ab88ff1382650c7152edfbc13ede6
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-23 11:37:31 +0000

    strip the script and move logic to a package 

  Change: 33446cd2f10522606421c5288539748c8970861a
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-23 10:50:30 +0000

    fixup++ 

  Change: 38ee78b77b6bf812e62c22e4f226ed1e2c57b683
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 13:52:27 +0000

    try and auto-sign git tags 

------------------------------------------
version 0.006 at 2014-06-20 17:36:24 +0000
------------------------------------------

  Change: dfb222f4e31d4aafd1f00594b84be0403d6a032f
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 13:36:24 +0000

    v0.006 

  Change: 28a557c6ccfa8d6fc103117b12ee833e40546443
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 13:34:43 +0000

    stub module 

  Change: 0e4b5e44f4e2a2cc6c01acb9356e4479ee0387d7
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 13:33:24 +0000

    Handle VERIFIED bugs as RESOLVED Cleaner verbose logging of only
    changes

    Not implemnted yet: bugs reopened should be moved back to Ready 

  Change: 2b0bb9b398a814f500b7d6aa516429f1d9554585
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 12:59:22 +0000

    When creating a card, set the title correctly [bug] 

  Change: 44a846b7d1cebbaaf4cde7979965e598d6dbbee5
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 12:58:49 +0000

    batch bug retrieval 

  Change: a91923d50a93e92f018b04a082ae8c744f8cee1b
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 12:20:57 +0000

    verbose-- 

  Change: e79af08ca82b734f34fa7f61004f799010d9eadf
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-20 12:18:35 +0000

    Add findbin to find itself 

------------------------------------------
version 0.005 at 2014-06-19 19:54:51 +0000
------------------------------------------

  Change: d7a225f81a8cf9da4c9b900cc08db88a3ba48081
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:54:51 +0000

    automate releases 

  Change: c142bf66c5ddd2615aada3b914d53c898933ea60
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:53:53 +0000

    Perl::Tidy 

------------------------------------------
version 0.004 at 2014-06-19 19:43:55 +0000
------------------------------------------

  Change: eb19278ff5ca9863e3cf03ea74bc098bb8528bd2
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:43:55 +0000

    v0.004 

  Change: e423aee7128e0c3394bb3f8c4641904eaca2ea6c
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:43:33 +0000

    use warnings 

  Change: a36af2bc308ba18eac7189d5894fdb1fa457f59a
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:43:24 +0000

    Add pre-requisites and meta.json 

  Change: 066b1a0a6797543c396c26b119d2752e0f2fb45b
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:25:56 +0000

    add critic tests 

  Change: 022f04975914e49e5c40cf649cc8af177a2c8f40
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:23:00 +0000

    dont ship sync.cfg 

  Change: d8c1384af04306eb25b35051289640d6f4f7bbca
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:12:07 +0000

    Git::NextVersion 

------------------------------------------
version 0.003 at 2014-06-19 19:08:29 +0000
------------------------------------------

  Change: 537409bc2beae615b8f193f85388d525de0c0414
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:08:29 +0000

    v0.003 

  Change: 2e241b45af2201ecbc59e0030fdeb0d03656c647
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:06:18 +0000

    Find bugs just marked with [kanban] 

  Change: 61b14b43d9021b68037480b85a266454878328f2
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 15:06:06 +0000

    return true 

------------------------------------------
version 0.002 at 2014-06-19 18:54:51 +0000
------------------------------------------

  Change: de735a8e23e36fb6f7f2d4b6076dfe5ae9a4635c
  Author: Philippe M. Chiasson <gozer@mozilla.com>
  Date : 2014-06-19 14:54:51 +0000

    v0.002 


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Bugzilla and Kanbanize sync tool",
  "AUTHOR" => "Philippe M. Chiasson <gozer\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Net-Bugzilla-Kanbanize",
  "EXE_FILES" => [
    "bin/kanbanize-bugzilla-sync"
  ],
  "LICENSE" => "mozilla",
  "NAME" => "Net::Bugzilla::Kanbanize",
  "PREREQ_PM" => {
    "AppConfig" => 0,
    "Data::Dumper" => 0,
    "File::HomeDir" => 0,
    "FindBin" => 0,
    "HTTP::Request" => 0,
    "JSON" => 0,
    "LWP::Simple" => 0,
    "LWP::UserAgent" => 0,
    "List::MoreUtils" => 0,
    "URI::Escape" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Module::Signature" => 0,
    "Socket" => 0
  },
  "VERSION" => "0.007",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "AppConfig" => 0,
  "Data::Dumper" => 0,
  "File::HomeDir" => 0,
  "FindBin" => 0,
  "HTTP::Request" => 0,
  "JSON" => 0,
  "LWP::Simple" => 0,
  "LWP::UserAgent" => 0,
  "List::MoreUtils" => 0,
  "Module::Signature" => 0,
  "Socket" => 0,
  "URI::Escape" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Net/Bugzilla/Kanbanize.pm
use strict;
use warnings;

package Net::Bugzilla::Kanbanize;
$Net::Bugzilla::Kanbanize::VERSION = '0.007';    # TRIAL
our $VERSION;

#ABSTRACT: Bugzilla and Kanbanize sync tool

use Data::Dumper;

use Net::Bugzilla::Kanbanize;

use LWP::Simple;
use JSON;

use LWP::UserAgent;
use File::HomeDir;

use HTTP::Request;
use URI::Escape;
use List::MoreUtils qw(uniq);

#XXX: https://bugzil.la/970457

sub new {
    my ( $class, $config ) = @_;

    my $self = bless { config => $config }, $class;

    return $self;
}

sub version {
    print STDERR "Version $VERSION\n";
}

#XXX: Wrong, need to be instance variables

my $APIKEY;
my $BOARD_ID;
my $BUGZILLA_TOKEN;
my $ua = LWP::UserAgent->new();

my $total;
my $count;
my $config;

sub run {
    my $self = shift;

    $config = $self->{config};

    $APIKEY = $config->kanbanize_apikey or die "Please configure an apikey";
    $BOARD_ID = $config->kanbanize_boardid
      or die "Please configure a kanbanize_boardid";
    $BUGZILLA_TOKEN = $config->bugzilla_token
      or die "Please configure a bugzilla_token";

    $ua->timeout(15);
    $ua->env_proxy;
    $ua->default_header( 'apikey' => $APIKEY );

    my %bugs;

    if (@ARGV) {
        fill_missing_bugs_info( \%bugs, @ARGV );
    }
    else {
        %bugs = get_bugs();
    }

    $count = scalar keys %bugs;

    print STDERR "Found a total of $count bugs\n";

    $total = 0;

    while ( my ( $bugid, $bug ) = each %bugs ) {
        sync_bug($bug);
    }

    return 1;
}

sub get_bugs {
    my $req =
      HTTP::Request->new( GET =>
"https://bugzilla.mozilla.org/rest/bug?token=$BUGZILLA_TOKEN&include_fields=id,status,whiteboard,summary,assigned_to&bug_status=UNCONFIRMED&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&component=WebOps%3A Bugzilla&component=WebOps%3A Community Platform&component=WebOps%3A Engagement&component=WebOps%3A IT-Managed Tools&component=WebOps%3A Labs&component=WebOps%3A Other&component=WebOps%3A Product Delivery&component=WebOps%3A SSL and Domain Names&product=Infrastructure %26 Operations"
      );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }

    my $data = decode_json( $res->decoded_content );

    my %bugs;

    foreach my $bug ( @{ $data->{bugs} } ) {
        $bugs{ $bug->{id} } = $bug;
    }

    my @marked = get_marked_bugs();
    foreach my $bug (@marked) {
        $bugs{ $bug->{id} } = $bug;
    }

    my @cards = get_bugs_from_all_cards();

    fill_missing_bugs_info( \%bugs, @cards );

    return %bugs;
}

sub fill_missing_bugs_info {
    my ( $bugs, @bugs ) = @_;

    my @missing_bugs;

    foreach my $bugid (@bugs) {
        if ( not exists $bugs->{$bugid} ) {
            push @missing_bugs, $bugid;
        }
    }

    my $missing_bugs_ids = join ",", sort @bugs;

    my $req =
      HTTP::Request->new( GET =>
"https://bugzilla.mozilla.org/rest/bug?token=$BUGZILLA_TOKEN&include_fields=id,status,whiteboard,summary,assigned_to&id=$missing_bugs_ids"
      );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }

    my $data = decode_json( $res->decoded_content );

    my @found_bugs = @{ $data->{bugs} };

    foreach my $bug ( sort @found_bugs ) {
        $bugs->{ $bug->{id} } = $bug;
    }

    return;
}

sub get_marked_bugs {
    my $req =
      HTTP::Request->new( GET =>
"https://bugzilla.mozilla.org/rest/bug?token=$BUGZILLA_TOKEN&include_fields=id,status,whiteboard,summary,assigned_to&status_whiteboard_type=allwordssubstr&query_format=advanced&status_whiteboard=[kanban]"
      );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }

    my $data = decode_json( $res->decoded_content );

    my @bugs = @{ $data->{bugs} };

    return @bugs;
}

sub get_bugs_from_all_cards {

    my $req =
      HTTP::Request->new( POST =>
"http://kanbanize.com/index.php/api/kanbanize/get_all_tasks/boardid/$BOARD_ID/format/json"
      );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }

    my $cards = decode_json( $res->decoded_content );

    my @bugs;
    foreach my $card (@$cards) {
        my $extlink = $card->{extlink};    # XXX: Smarter parsing
        if ( $extlink =~ /(\d+)$/ ) {
            my $bugid = $1;
            push @bugs, $bugid;
        }
    }

    return @bugs;
}

sub sync_bug {
    my $bug = shift;

    #    print STDERR "Bugid: $bug->{id}\n" if $config->verbose;

    $total++;

    if ( not defined $bug ) {
        print STDERR "[$total/$count] No info for bug $bug->{id}\n";
        return;
    }

    if ( $bug->{error} ) {
        print STDERR
          "[$total/$count] No info for bug $bug->{id} (Private bug?)\n";
        return;
    }

    my $summary    = $bug->{summary};
    my $whiteboard = $bug->{whiteboard};

    my $card = parse_whiteboard($whiteboard);

    my $status = "";
    if ( not defined $card ) {
        $card = create_card($bug);

        if ( not $card ) {
            warn "Failed to create card for bug $bug->{id}";
            return;
        }

        update_whiteboard( $bug->{id}, $card->{taskid}, $whiteboard );

        $status .= "[card created]";
    }

    $card = retrieve_card( $card->{taskid} );

    my $cardid = $card->{taskid};

    my @changes = sync_card( $card, $bug );
    if (@changes) {
        $status .= " [synced]";
    }

    if ( $status ne "" or $config->verbose ) {
        printf STDERR "[%4d/%4d] Card %4d - Bug %8d - $summary $status\n",
          $total, $count, $cardid, $bug->{id};
    }

    if (@changes) {
        foreach my $change (@changes) {
            printf STDERR "[%4d/%4d] Card %4d - Bug %8d - $summary ** %s **\n",
              $total, $count, $cardid, $bug->{id}, $change;
        }
    }
}

sub retrieve_card {
    my $card_id = shift;

    my $req =
      HTTP::Request->new( POST =>
"http://kanbanize.com/index.php/api/kanbanize/get_task_details/boardid/$BOARD_ID/taskid/$card_id/format/json"
      );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }

    my $card = decode_json( $res->decoded_content );

    return $card;
}

sub sync_bugzilla {

}

sub sync_card {
    my ( $card, $bug ) = @_;

    my @updated;

    # Check Assignee
    my $bug_assigned  = $bug->{assigned_to};
    my $card_assigned = $card->{assignee};

    if (   defined $card_assigned
        && $card_assigned ne "None"
        && $bug_assigned =~ m/\@.*\.bugs$/ )
    {
        push @updated, "Update bug $bug->{id} assigned to $card_assigned";
        update_bug_assigned( $bug, $card_assigned );
    }
    elsif ($bug_assigned !~ m[^$card_assigned@]
        && $bug_assigned !~ m/\@.*\.bugs$/ )
    {
        push @updated, "Update card assigned to $bug_assigned";

        #print STDERR
        # "bug_asigned: $bug_assigned card_assigned: $card_assigned\n";
        update_card_assigned( $card, $bug_assigned );
    }

    #Check summary (XXX: Formatting assumption here)
    my $bug_summary  = "$bug->{id} - $bug->{summary}";
    my $card_summary = $card->{title};

    if ( $bug_summary ne $card_summary ) {
        update_card_summary( $card, $bug_summary );
        push @updated, "Updated card summary";
    }

    # Check status
    my $bug_status  = $bug->{status};
    my $card_status = $card->{columnname};

    # Close card on bug completion

   #warn "[$bug->{id}] bug: $bug_status card: $card_status" if $config->verbose;

    if ( ( $bug_status eq "RESOLVED" or $bug_status eq "VERIFIED" )
        and $card_status ne "Done" )
    {
        complete_card($card);
        push @updated, "Card completed";
    }

    # XXX: Should we close bug on card completion?
    if ( ( $bug_status ne "RESOLVED" and $bug_status ne "VERIFIED" )
        and $card_status eq "Done" )
    {
        if ( $bug_status eq "REOPENED" ) {
            reopen_card($card);

            #$updated++;
        }
        else {
            warn
"Bug $bug->{id} is not RESOLVED ($bug_status) but card $card->{taskid} says $card_status";
        }
    }

    # Check extlink
    my $bug_link = "https://bugzilla.mozilla.org/show_bug.cgi?id=$bug->{id}";

    if ( $card->{extlink} ne $bug_link ) {
        update_card_extlink( $card, $bug_link );
        push @updated, "Updated external link to bugzilla";
    }

    return @updated;
}

sub reopen_card {
    my $card = shift;

    warn
"[notimplemented] Should be reopening card $card->{taskid} and moving back to ready";

    return;
}

sub complete_card {
    my $card = shift;

    my $taskid = $card->{taskid};

    my $data = {
        boardid => $BOARD_ID,
        taskid  => $taskid,
        column  => 'Done',
    };

    my $req =
      HTTP::Request->new( POST =>
          "http://kanbanize.com/index.php/api/kanbanize/move_task/format/json"
      );

    $req->content( encode_json($data) );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }
}

sub update_card_extlink {
    my ( $card, $extlink ) = @_;

    my $taskid = $card->{taskid};

    my $data = {
        boardid => $BOARD_ID,
        taskid  => $taskid,
        extlink => $extlink,
    };

    my $req =
      HTTP::Request->new( POST =>
          "http://kanbanize.com/index.php/api/kanbanize/edit_task/format/json"
      );

    $req->content( encode_json($data) );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }
}

sub update_bug_assigned {
    my ( $bug, $assigned ) = @_;

    $assigned .= '@mozilla.com';

    my $bugid = $bug->{id};

    my $req =
      HTTP::Request->new(
        PUT => "https://bugzilla.mozilla.org/rest/bug/$bugid" );

    $req->content("assigned_to=$assigned&token=$BUGZILLA_TOKEN");

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }
}

sub update_card_summary {
    my ( $card, $bug_summary ) = @_;

    my $taskid = $card->{taskid};

    my $data = {
        boardid => $BOARD_ID,
        taskid  => $taskid,
        title   => $bug_summary,
    };

    my $req =
      HTTP::Request->new( POST =>
          "http://kanbanize.com/index.php/api/kanbanize/edit_task/format/json"
      );

    $req->content( encode_json($data) );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);    #$res->status_line;
    }
}

sub update_card_assigned {
    my ( $card, $bug_assigned ) = @_;

    my $taskid = $card->{taskid};
    ( my $assignee = $bug_assigned ) =~ s/\@.*//;

    my $req =
      HTTP::Request->new( POST =>
"http://kanbanize.com/index.php/api/kanbanize/edit_task/format/json/boardid/$BOARD_ID/taskid/$taskid/assignee/$assignee"
      );

    $req->content("[]");

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die $res->status_line;
    }

}

sub update_whiteboard {
    my ( $bugid, $cardid, $whiteboard ) = @_;

    my $req =
      HTTP::Request->new(
        PUT => "https://bugzilla.mozilla.org/rest/bug/$bugid" );

    if ( $whiteboard =~ m/\[kanban\]/ ) {
        $whiteboard =~ s/\[kanban\]//;
    }

    $whiteboard =
      "[kanban:https://kanbanize.com/ctrl_board/$BOARD_ID/$cardid] $whiteboard";

    $req->content("whiteboard=$whiteboard&token=$BUGZILLA_TOKEN");

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die $res->status_line;
    }

}

#XXX: https://bugzil.la/970457
sub create_card {
    my $bug = shift;

    my $data = {
        'title'   => "$bug->{id} - $bug->{summary}",
        'extlink' => "https://bugzilla.mozilla.org/show_bug.cgi?id=$bug->{id}",
        'boardid' => $BOARD_ID,
    };

    my $req =
      HTTP::Request->new( POST =>
"http://kanbanize.com/index.php/api/kanbanize/create_new_task/format/json"
      );

    $req->content( encode_json($data) );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        warn "can't create card:" . $res->status_line;
        die Dumper($res);
        return;
    }

    my $card = decode_json( $res->decoded_content );

    $card->{taskid} = $card->{id};

    move_card( $card, 'Pending Triage' );

    return $card;
}

sub move_card {
    my ( $card, $lane ) = @_;

    my $data = {
        boardid => $BOARD_ID,
        taskid  => $card->{taskid},
        column  => 'Backlog',
        lane    => $lane,
    };

    my $req =
      HTTP::Request->new( POST =>
          "http://kanbanize.com/index.php/api/kanbanize/move_task/format/json"
      );

    $req->content( encode_json($data) );

    my $res = $ua->request($req);

    if ( !$res->is_success ) {
        die Dumper($res);
    }

}

sub get_bug_info {
    my $bugid = shift;
    my $data =
      get("https://bugzilla.mozilla.org/rest/bug/$bugid?token=$BUGZILLA_TOKEN");

    if ( not $data ) {
        warn "Failed getting Bug info for Bug $bugid from bugzilla\n";
        return { id => $bugid, error => "No Data" };
    }

    print STDERR "Retrieving info for Bug $bugid from bugzilla\n"
      if $config->verbose;

    $data = decode_json($data);

    return $data->{bugs}[0];
}

sub parse_whiteboard {
    my $whiteboard = shift;

    my $card;

    if ( $whiteboard =~
        m{\[kanban:https://kanbanize.com/ctrl_board/(\d+)/(\d+)\]} )
    {
        my $boardid = $1;
        my $cardid  = $2;

        $card = { taskid => $cardid };
    }

    return $card;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Net::Bugzilla::Kanbanize - Bugzilla and Kanbanize sync tool

=head1 VERSION

version 0.007

=head1 SYNOPSIS

Kanbanize Bugzilla Sync Tool

=head2 version

prints current version to STDERR

=head1 METHODS

=head2 new

This method does something experimental.

=head2 version

This method returns a reason.

=head1 AUTHOR

Philippe M. Chiasson <gozer@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Philippe M. Chiasson.

This is free software, licensed under:

  Mozilla Public License Version 2.0

=cut
